File: so101_ros2/dapier_localization_core/src/localization_runtime.cpp
// SPDX-License-Identifier: Apache-2.0

#include "dapier_localization_core/localization_runtime.hpp"

#include "dapier_localization_core/generated/localization_runtime_contract.hpp"

#include <algorithm>
#include <cctype>
#include <cmath>
#include <cstddef>
#include <cstdint>
#include <limits>
#include <string>
#include <utility>

namespace dapier_localization_core
{
namespace
{

static_assert(generated::kAllowedTrackingStates.size() == 5);
static_assert(generated::kAllowedTrackingStates[0] == "initializing");
static_assert(generated::kAllowedTrackingStates[1] == "tracking");
static_assert(generated::kAllowedTrackingStates[2] == "recently_lost");
static_assert(generated::kAllowedTrackingStates[3] == "lost");
static_assert(generated::kAllowedTrackingStates[4] == "relocalized");
static_assert(!generated::kLocalizationMayAuthorizeHardware);
static_assert(!generated::kSimulatorTruthAllowedForRuntime);

LocalizationValidation reject(
  LocalizationRejection rejection,
  std::string reason)
{
  return LocalizationValidation{false, rejection, std::move(reason), 0};
}

MotionAssessment motion(MotionDecision decision, std::string reason)
{
  return MotionAssessment{decision, std::move(reason)};
}

bool is_blank(const std::string & value)
{
  return value.empty() || std::all_of(
    value.begin(), value.end(), [](unsigned char character) {
      return std::isspace(character) != 0;
    });
}

bool invalid_identifier(const std::string & value)
{
  return is_blank(value) || value.size() > generated::kMaxIdentifierLength;
}

template<std::size_t Size>
bool contains_non_finite(const std::array<double, Size> & values)
{
  return std::any_of(values.begin(), values.end(), [](double value) {
    return !std::isfinite(value);
  });
}

bool covariance_is_symmetric(const std::array<double, 36> & covariance)
{
  constexpr double tolerance = 1e-8;
  for (std::size_t row = 0; row < 6; ++row) {
    for (std::size_t column = row + 1; column < 6; ++column) {
      const auto forward = covariance[row * 6 + column];
      const auto reverse = covariance[column * 6 + row];
      if (std::abs(forward - reverse) > tolerance) {
        return false;
      }
    }
  }
  return true;
}

bool gate_config_is_valid(const MotionGateConfig & config)
{
  return std::isfinite(config.minimum_confidence) &&
         config.minimum_confidence >= 0.0 &&
         config.minimum_confidence <= 1.0 &&
         config.minimum_tracked_features > 0 &&
         std::isfinite(config.maximum_mean_reprojection_error_px) &&
         config.maximum_mean_reprojection_error_px > 0.0 &&
         std::isfinite(config.maximum_position_variance_m2) &&
         config.maximum_position_variance_m2 > 0.0 &&
         std::isfinite(config.maximum_rotation_variance_rad2) &&
         config.maximum_rotation_variance_rad2 > 0.0;
}

}  // namespace

LocalizationValidation validate_localization_estimate(
  const LocalizationEstimate & estimate,
  std::int64_t receiver_monotonic_ns,
  std::optional<std::uint64_t> last_accepted_sequence,
  std::optional<std::uint64_t> expected_reset_generation)
{
  if (estimate.schema_version != generated::kLocalizationSchemaVersion) {
    return reject(
      LocalizationRejection::kSchemaMismatch,
      "schema version does not match the generated localization boundary");
  }
  if (last_accepted_sequence && estimate.sequence <= *last_accepted_sequence) {
    return reject(
      LocalizationRejection::kSequenceRejected,
      "sequence must be greater than the last accepted sequence");
  }
  if (expected_reset_generation &&
    estimate.reset_generation != *expected_reset_generation)
  {
    return reject(
      LocalizationRejection::kGenerationMismatch,
      "localization reset generation differs from the active plan generation");
  }
  if (receiver_monotonic_ns < 0 || estimate.source_timestamp_ns < 0) {
    return reject(
      LocalizationRejection::kInvalidTimestamp,
      "source and receiver timestamps must be non-negative");
  }
  if (estimate.ttl_ns <= 0 ||
    estimate.ttl_ns > generated::kMaxEstimateTtlNs)
  {
    return reject(
      LocalizationRejection::kInvalidTtl,
      "ttl is outside the localization contract maximum");
  }
  if (invalid_identifier(estimate.parent_frame) ||
    invalid_identifier(estimate.child_frame) ||
    invalid_identifier(estimate.map_id) ||
    invalid_identifier(estimate.session_id))
  {
    return reject(
      LocalizationRejection::kInvalidIdentifier,
      "frame and map/session identifiers must be non-empty and bounded");
  }
  if (estimate.parent_frame == estimate.child_frame) {
    return reject(
      LocalizationRejection::kInvalidFramePair,
      "parent and child frames must differ");
  }
  switch (estimate.tracking_state) {
    case TrackingState::kInitializing:
    case TrackingState::kTracking:
    case TrackingState::kRecentlyLost:
    case TrackingState::kLost:
    case TrackingState::kRelocalized:
      break;
    default:
      return reject(
        LocalizationRejection::kUnknownTrackingState,
        "tracking state is not part of the generated localization boundary");
  }
  if (contains_non_finite(estimate.position_m) ||
    contains_non_finite(estimate.orientation_xyzw) ||
    contains_non_finite(estimate.covariance) ||
    !std::isfinite(estimate.mean_reprojection_error_px) ||
    !std::isfinite(estimate.confidence))
  {
    return reject(
      LocalizationRejection::kNonFiniteValue,
      "pose, covariance, and quality fields must be finite");
  }

  double quaternion_norm_squared = 0.0;
  for (double value : estimate.orientation_xyzw) {
    quaternion_norm_squared += value * value;
  }
  if (!std::isfinite(quaternion_norm_squared) ||
    std::abs(quaternion_norm_squared - 1.0) > 1e-3)
  {
    return reject(
      LocalizationRejection::kInvalidQuaternion,
      "orientation quaternion must be normalized");
  }

  constexpr std::array<std::size_t, 6> diagonal_indices{0, 7, 14, 21, 28, 35};
  for (std::size_t index : diagonal_indices) {
    if (estimate.covariance[index] < 0.0) {
      return reject(
        LocalizationRejection::kInvalidCovariance,
        "covariance diagonal entries must be non-negative");
    }
  }
  if (!covariance_is_symmetric(estimate.covariance)) {
    return reject(
      LocalizationRejection::kInvalidCovariance,
      "covariance must be symmetric");
  }
  if (estimate.mean_reprojection_error_px < 0.0 ||
    estimate.confidence < 0.0 || estimate.confidence > 1.0)
  {
    return reject(
      LocalizationRejection::kInvalidQuality,
      "quality metrics are outside their valid ranges");
  }
  if (estimate.simulator_truth_used) {
    return reject(
      LocalizationRejection::kPrivilegedRuntimeState,
      "runtime localization must not depend on simulator truth");
  }
  if (estimate.control_authorized) {
    return reject(
      LocalizationRejection::kHardwareAuthorityViolation,
      "localization estimates cannot authorize hardware control");
  }
  if (estimate.ttl_ns >
    std::numeric_limits<std::int64_t>::max() - receiver_monotonic_ns)
  {
    return reject(
      LocalizationRejection::kExpiryOverflow,
      "receiver-local expiry would overflow int64");
  }

  return LocalizationValidation{
    true,
    LocalizationRejection::kNone,
    "accepted by the localization boundary; motion-quality gates still apply",
    receiver_monotonic_ns + estimate.ttl_ns,
  };
}

MotionAssessment assess_localization_for_motion(
  const LocalizationEstimate & estimate,
  const LocalizationValidation & validation,
  const MotionGateConfig & config)
{
  if (!validation.accepted) {
    return motion(
      MotionDecision::kReject,
      "invalid localization estimate cannot be used for motion");
  }
  if (!gate_config_is_valid(config)) {
    return motion(
      MotionDecision::kReject,
      "motion gate configuration is invalid");
  }

  switch (estimate.tracking_state) {
    case TrackingState::kTracking:
      break;
    case TrackingState::kRelocalized:
      return motion(
        MotionDecision::kHold,
        "relocalized pose requires target invalidation and replanning");
    case TrackingState::kInitializing:
      return motion(MotionDecision::kHold, "localization is initializing");
    case TrackingState::kRecentlyLost:
      return motion(MotionDecision::kHold, "localization is recently lost");
    case TrackingState::kLost:
      return motion(MotionDecision::kHold, "localization is lost");
    default:
      return motion(MotionDecision::kReject, "tracking state is invalid");
  }

  if (estimate.confidence < config.minimum_confidence) {
    return motion(MotionDecision::kHold, "localization confidence is too low");
  }
  if (estimate.tracked_features < config.minimum_tracked_features) {
    return motion(MotionDecision::kHold, "tracked feature count is too low");
  }
  if (estimate.mean_reprojection_error_px >
    config.maximum_mean_reprojection_error_px)
  {
    return motion(MotionDecision::kHold, "reprojection error is too high");
  }

  for (std::size_t index : {std::size_t{0}, std::size_t{7}, std::size_t{14}}) {
    if (estimate.covariance[index] > config.maximum_position_variance_m2) {
      return motion(MotionDecision::kHold, "position covariance is too high");
    }
  }
  for (std::size_t index : {std::size_t{21}, std::size_t{28}, std::size_t{35}}) {
    if (estimate.covariance[index] > config.maximum_rotation_variance_rad2) {
      return motion(MotionDecision::kHold, "rotation covariance is too high");
    }
  }

  return motion(MotionDecision::kProceed, "localization is suitable for motion");
}

}  // namespace dapier_localization_core

File: 2ARM_ROBOT/research/src/dapier_research/localization_runtime.py
"""Portable localization estimate validation for replay and policy planning.

This module mirrors the language-neutral ``localization_runtime_v1`` contract.
Source timestamps are trace metadata only. Freshness is always calculated from
receiver-local monotonic time by the consumer.
"""

from __future__ import annotations

from dataclasses import dataclass
import json
import math
from pathlib import Path
from typing import Any, Mapping, Sequence


LOCALIZATION_SCHEMA_VERSION = "dapier.localization-estimate.v1"
MAX_ESTIMATE_TTL_NS = 250_000_000
MAX_IDENTIFIER_LENGTH = 128
TRACKING_STATES = (
    "initializing",
    "tracking",
    "recently_lost",
    "lost",
    "relocalized",
)
MOTION_DECISIONS = ("proceed", "hold", "reject")

_REQUIRED_KEYS = frozenset(
    {
        "schema_version",
        "sequence",
        "source_timestamp_ns",
        "ttl_ns",
        "parent_frame",
        "child_frame",
        "map_id",
        "session_id",
        "reset_generation",
        "tracking_state",
        "position_m",
        "orientation_xyzw",
        "covariance",
        "tracked_features",
        "mean_reprojection_error_px",
        "confidence",
        "simulator_truth_used",
        "control_authorized",
    }
)


class LocalizationContractError(ValueError):
    """Raised when a localization estimate violates the public contract."""


@dataclass(frozen=True)
class LocalizationEstimate:
    schema_version: str
    sequence: int
    source_timestamp_ns: int
    ttl_ns: int
    parent_frame: str
    child_frame: str
    map_id: str
    session_id: str
    reset_generation: int
    tracking_state: str
    position_m: tuple[float, float, float]
    orientation_xyzw: tuple[float, float, float, float]
    covariance: tuple[float, ...]
    tracked_features: int
    mean_reprojection_error_px: float
    confidence: float
    simulator_truth_used: bool
    control_authorized: bool

    def as_dict(self) -> dict[str, Any]:
        return {
            "schema_version": self.schema_version,
            "sequence": self.sequence,
            "source_timestamp_ns": self.source_timestamp_ns,
            "ttl_ns": self.ttl_ns,
            "parent_frame": self.parent_frame,
            "child_frame": self.child_frame,
            "map_id": self.map_id,
            "session_id": self.session_id,
            "reset_generation": self.reset_generation,
            "tracking_state": self.tracking_state,
            "position_m": list(self.position_m),
            "orientation_xyzw": list(self.orientation_xyzw),
            "covariance": list(self.covariance),
            "tracked_features": self.tracked_features,
            "mean_reprojection_error_px": self.mean_reprojection_error_px,
            "confidence": self.confidence,
            "simulator_truth_used": self.simulator_truth_used,
            "control_authorized": self.control_authorized,
        }


@dataclass(frozen=True)
class LocalizationValidation:
    accepted: bool
    reason: str
    accepted_until_monotonic_ns: int | None
    estimate: LocalizationEstimate | None = None


@dataclass(frozen=True)
class MotionGateConfig:
    minimum_confidence: float = 0.5
    minimum_tracked_features: int = 20
    maximum_mean_reprojection_error_px: float = 3.0
    maximum_position_variance_m2: float = 0.01
    maximum_rotation_variance_rad2: float = 0.04

    def validate(self) -> None:
        numeric = (
            self.minimum_confidence,
            self.maximum_mean_reprojection_error_px,
            self.maximum_position_variance_m2,
            self.maximum_rotation_variance_rad2,
        )
        if not all(math.isfinite(value) for value in numeric):
            raise LocalizationContractError("motion gate values must be finite")
        if not 0.0 <= self.minimum_confidence <= 1.0:
            raise LocalizationContractError("minimum_confidence must be in [0, 1]")
        if isinstance(self.minimum_tracked_features, bool) or self.minimum_tracked_features <= 0:
            raise LocalizationContractError("minimum_tracked_features must be positive")
        if self.maximum_mean_reprojection_error_px <= 0.0:
            raise LocalizationContractError(
                "maximum_mean_reprojection_error_px must be positive"
            )
        if self.maximum_position_variance_m2 <= 0.0:
            raise LocalizationContractError(
                "maximum_position_variance_m2 must be positive"
            )
        if self.maximum_rotation_variance_rad2 <= 0.0:
            raise LocalizationContractError(
                "maximum_rotation_variance_rad2 must be positive"
            )


@dataclass(frozen=True)
class MotionAssessment:
    decision: str
    reason: str


def _text(value: object, name: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise LocalizationContractError(f"{name} must be non-empty text")
    if len(value) > MAX_IDENTIFIER_LENGTH:
        raise LocalizationContractError(f"{name} exceeds {MAX_IDENTIFIER_LENGTH} characters")
    return value


def _integer(value: object, name: str, *, positive: bool = False) -> int:
    if isinstance(value, bool) or not isinstance(value, int):
        raise LocalizationContractError(f"{name} must be an integer")
    if value < 0 or (positive and value <= 0):
        qualifier = "positive" if positive else "non-negative"
        raise LocalizationContractError(f"{name} must be {qualifier}")
    return value


def _number(value: object, name: str) -> float:
    if isinstance(value, bool) or not isinstance(value, (int, float)):
        raise LocalizationContractError(f"{name} must be numeric")
    result = float(value)
    if not math.isfinite(result):
        raise LocalizationContractError(f"{name} must be finite")
    return result


def _vector(value: object, name: str, length: int) -> tuple[float, ...]:
    if not isinstance(value, Sequence) or isinstance(value, (str, bytes)):
        raise LocalizationContractError(f"{name} must be an array")
    if len(value) != length:
        raise LocalizationContractError(f"{name} must contain {length} values")
    return tuple(_number(item, f"{name}[{index}]") for index, item in enumerate(value))


def estimate_from_mapping(value: Mapping[str, object]) -> LocalizationEstimate:
    if not isinstance(value, Mapping):
        raise LocalizationContractError("localization estimate must be an object")
    keys = frozenset(value)
    if keys != _REQUIRED_KEYS:
        raise LocalizationContractError(
            "localization estimate keys differ; "
            f"missing={sorted(_REQUIRED_KEYS - keys)}, extra={sorted(keys - _REQUIRED_KEYS)}"
        )
    estimate = LocalizationEstimate(
        schema_version=_text(value["schema_version"], "schema_version"),
        sequence=_integer(value["sequence"], "sequence"),
        source_timestamp_ns=_integer(
            value["source_timestamp_ns"], "source_timestamp_ns"
        ),
        ttl_ns=_integer(value["ttl_ns"], "ttl_ns", positive=True),
        parent_frame=_text(value["parent_frame"], "parent_frame"),
        child_frame=_text(value["child_frame"], "child_frame"),
        map_id=_text(value["map_id"], "map_id"),
        session_id=_text(value["session_id"], "session_id"),
        reset_generation=_integer(value["reset_generation"], "reset_generation"),
        tracking_state=_text(value["tracking_state"], "tracking_state"),
        position_m=_vector(value["position_m"], "position_m", 3),
        orientation_xyzw=_vector(
            value["orientation_xyzw"], "orientation_xyzw", 4
        ),
        covariance=_vector(value["covariance"], "covariance", 36),
        tracked_features=_integer(value["tracked_features"], "tracked_features"),
        mean_reprojection_error_px=_number(
            value["mean_reprojection_error_px"], "mean_reprojection_error_px"
        ),
        confidence=_number(value["confidence"], "confidence"),
        simulator_truth_used=value["simulator_truth_used"],
        control_authorized=value["control_authorized"],
    )
    if not isinstance(estimate.simulator_truth_used, bool):
        raise LocalizationContractError("simulator_truth_used must be boolean")
    if not isinstance(estimate.control_authorized, bool):
        raise LocalizationContractError("control_authorized must be boolean")
    return estimate


def validate_localization_estimate(
    value: Mapping[str, object] | LocalizationEstimate,
    *,
    receiver_monotonic_ns: int,
    last_accepted_sequence: int | None = None,
    expected_reset_generation: int | None = None,
) -> LocalizationValidation:
    try:
        estimate = (
            value
            if isinstance(value, LocalizationEstimate)
            else estimate_from_mapping(value)
        )
        receiver = _integer(receiver_monotonic_ns, "receiver_monotonic_ns")
        if estimate.schema_version != LOCALIZATION_SCHEMA_VERSION:
            raise LocalizationContractError("unsupported localization schema_version")
        if estimate.ttl_ns > MAX_ESTIMATE_TTL_NS:
            raise LocalizationContractError("ttl exceeds the localization contract")
        if last_accepted_sequence is not None:
            previous = _integer(last_accepted_sequence, "last_accepted_sequence")
            if estimate.sequence <= previous:
                raise LocalizationContractError(
                    "sequence must be greater than the last accepted sequence"
                )
        if expected_reset_generation is not None:
            expected = _integer(
                expected_reset_generation, "expected_reset_generation"
            )
            if estimate.reset_generation != expected:
                raise LocalizationContractError(
                    "reset_generation differs from the active plan generation"
                )
        if estimate.parent_frame == estimate.child_frame:
            raise LocalizationContractError("parent_frame and child_frame must differ")
        if estimate.tracking_state not in TRACKING_STATES:
            raise LocalizationContractError("tracking_state is not recognized")
        quaternion_norm = math.sqrt(
            sum(value * value for value in estimate.orientation_xyzw)
        )
        if abs(quaternion_norm - 1.0) > 1e-3:
            raise LocalizationContractError("orientation quaternion must be normalized")
        for index in (0, 7, 14, 21, 28, 35):
            if estimate.covariance[index] < 0.0:
                raise LocalizationContractError(
                    "covariance diagonal entries must be non-negative"
                )
        for row in range(6):
            for column in range(row + 1, 6):
                forward = estimate.covariance[row * 6 + column]
                reverse = estimate.covariance[column * 6 + row]
                if abs(forward - reverse) > 1e-8:
                    raise LocalizationContractError("covariance must be symmetric")
        if estimate.mean_reprojection_error_px < 0.0:
            raise LocalizationContractError(
                "mean_reprojection_error_px must be non-negative"
            )
        if not 0.0 <= estimate.confidence <= 1.0:
            raise LocalizationContractError("confidence must be in [0, 1]")
        if estimate.simulator_truth_used:
            raise LocalizationContractError(
                "runtime localization must not depend on simulator truth"
            )
        if estimate.control_authorized:
            raise LocalizationContractError(
                "localization cannot authorize hardware control"
            )
        if estimate.ttl_ns > (2**63 - 1) - receiver:
            raise LocalizationContractError("receiver-local expiry would overflow int64")
    except LocalizationContractError as error:
        return LocalizationValidation(False, str(error), None, None)
    return LocalizationValidation(
        True,
        "accepted by the localization boundary; quality gates still apply",
        receiver + estimate.ttl_ns,
        estimate,
    )


def assess_localization_for_motion(
    estimate: LocalizationEstimate,
    validation: LocalizationValidation,
    *,
    config: MotionGateConfig | None = None,
) -> MotionAssessment:
    if not validation.accepted:
        return MotionAssessment("reject", "invalid localization estimate")
    resolved = config or MotionGateConfig()
    try:
        resolved.validate()
    except LocalizationContractError as error:
        return MotionAssessment("reject", str(error))
    if estimate.tracking_state == "relocalized":
        return MotionAssessment(
            "hold", "relocalized pose requires target invalidation and replanning"
        )
    if estimate.tracking_state == "initializing":
        return MotionAssessment("hold", "localization is initializing")
    if estimate.tracking_state == "recently_lost":
        return MotionAssessment("hold", "localization is recently lost")
    if estimate.tracking_state == "lost":
        return MotionAssessment("hold", "localization is lost")
    if estimate.tracking_state != "tracking":
        return MotionAssessment("reject", "tracking state is invalid")
    if estimate.confidence < resolved.minimum_confidence:
        return MotionAssessment("hold", "localization confidence is too low")
    if estimate.tracked_features < resolved.minimum_tracked_features:
        return MotionAssessment("hold", "tracked feature count is too low")
    if (
        estimate.mean_reprojection_error_px
        > resolved.maximum_mean_reprojection_error_px
    ):
        return MotionAssessment("hold", "reprojection error is too high")
    if any(
        estimate.covariance[index] > resolved.maximum_position_variance_m2
        for index in (0, 7, 14)
    ):
        return MotionAssessment("hold", "position covariance is too high")
    if any(
        estimate.covariance[index] > resolved.maximum_rotation_variance_rad2
        for index in (21, 28, 35)
    ):
        return MotionAssessment("hold", "rotation covariance is too high")
    return MotionAssessment("proceed", "localization is suitable for motion")


def load_contract(path: str | Path) -> Mapping[str, Any]:
    contract_path = Path(path)
    try:
        value = json.loads(contract_path.read_text(encoding="utf-8"))
    except FileNotFoundError as error:
        raise LocalizationContractError(f"contract not found: {contract_path}") from error
    except json.JSONDecodeError as error:
        raise LocalizationContractError(f"invalid contract JSON: {error}") from error
    if not isinstance(value, Mapping):
        raise LocalizationContractError("contract must be a JSON object")
    return value


def verify_python_contract_matches_json(path: str | Path) -> None:
    contract = load_contract(path)
    expected = {
        "schema_version": LOCALIZATION_SCHEMA_VERSION,
        "max_estimate_ttl_ns": MAX_ESTIMATE_TTL_NS,
        "max_identifier_length": MAX_IDENTIFIER_LENGTH,
        "allowed_tracking_states": list(TRACKING_STATES),
        "localization_may_authorize_hardware": False,
        "simulator_truth_allowed_for_runtime": False,
    }
    for key, value in expected.items():
        if contract.get(key) != value:
            raise LocalizationContractError(
                f"Python localization constant differs from contract: {key}"
            )


__all__ = [
    "LOCALIZATION_SCHEMA_VERSION",
    "MAX_ESTIMATE_TTL_NS",
    "MAX_IDENTIFIER_LENGTH",
    "MOTION_DECISIONS",
    "TRACKING_STATES",
    "LocalizationContractError",
    "LocalizationEstimate",
    "LocalizationValidation",
    "MotionAssessment",
    "MotionGateConfig",
    "assess_localization_for_motion",
    "estimate_from_mapping",
    "load_contract",
    "validate_localization_estimate",
    "verify_python_contract_matches_json",
]
